src/java/org/jivesoftware/openfire/plugin/rest/controller/SystemController.java

```diff
@@ -0,0 +1,273 @@
+     * @param systemProperty the system property
+     * @throws ServiceException the service exception
+     */
+    public void updateSystemProperty(String propertyKey, org.jivesoftware.openfire.plugin.rest.entity.SystemProperty systemProperty) throws ServiceException {
+        if(JiveGlobals.getProperty(propertyKey) != null) {
+            if(systemProperty.getKey().equals(propertyKey)) {
+                JiveGlobals.setProperty(propertyKey, systemProperty.getValue());
+            } else {
+                throw new ServiceException("Path property name and entity property name doesn't match", propertyKey, ExceptionType.ILLEGAL_ARGUMENT_EXCEPTION,
+                    Response.Status.BAD_REQUEST);
+            }
+        } else {
+            throw new ServiceException("Could not find property for update", systemProperty.getKey(), ExceptionType.PROPERTY_NOT_FOUND,
+                Response.Status.NOT_FOUND);
+        }
+    }
+
+    /**
+     * Determines if there are threads that are deadlocked.
+     *
+     * @return true if deadlocked threads are found, otherwise false.
+     */
+    public boolean hasDeadlock() {
+        try {
+            return ManagementFactory.getThreadMXBean().findDeadlockedThreads() != null;
+        } catch (Throwable t) {
+            LOG.warn("Unable to determine if there is a deadlock.", t);
+            return false;
+        }
+    }
+
+    /**
+     * Determines if there are any system properties that require a restart.
+     *
+     * @return true if at least one system property requires a restart, otherwise false.
+     */
+    public boolean hasSystemPropertyRequiringRestart() {
+        try {
+            SystemProperty.getProperties().stream().filter(SystemProperty::isRestartRequired).forEach(p -> LOG.info("Requires restart: {}. Current value: {}. Initial value: {}", p.getKey(), p.getValue(), p.getDisplayValue()));
+            return SystemProperty.getProperties().stream()
+                .filter(systemProperty -> !systemProperty.getKey().equals("xmpp.domain")) // xmpp.domain can report a false positive. See OF-2399
+                .anyMatch(SystemProperty::isRestartRequired);
+        } catch (Throwable t) {
+            LOG.warn("Unable to determine if there are any system properties that require a restart.", t);
+            return false;
+        }
+    }
+
+    /**
+     * Determine if the core Openfire service is started.
+     *
+     * @return true if Openfire has started, otherwise false.
+     */
+    public boolean isStarted() {
+        return XMPPServer.getInstance().isStarted();
+    }
+
+    /**
+     * Determine if clustering has been fully started, if clustering is enabled.
+     *
+     * When clustering is not enabled, this method will always return 'true'.
+     *
+     * @return true if Openfire has started, otherwise false.
+     */
+    public boolean hasClusteringStartedWhenEnabled() {
+        if (!ClusterManager.isClusteringEnabled()) {
```

Comment:
Is "enabled" enough for this readiness check? Does it need to be settled?

Reply:
This uses the same reasoning as earlier. You are right that we need to wait for clustering to be settled, but that's done in the next few lines. If clustering is _disabled_, then the check should report successful (as Openfire is 'ready to be used' from this perspective).